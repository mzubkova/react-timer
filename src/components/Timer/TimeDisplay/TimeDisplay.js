import React, { useState, useEffect, useContext } from "react";
import PropTypes from "prop-types";

import { ThemeContext } from "../../Context/ThemeContext";

import "../Timer.scss";

const TimeDisplay = ({ initialTime, step = 1000, isActive }) => {
  const [time, setTime] = useState(initialTime);

  const { theme } = useContext(ThemeContext);

  useEffect(() => {
    setTime(initialTime);
  }, [initialTime]);

  useEffect(() => {
    let interval = null;
    if (isActive && time > 0) {
      interval = setInterval(() => {
        setTime((time) => time - 1);
      }, step);
    } else if (!isActive && time !== 0) {
      clearInterval(interval);
    }
    return () => clearInterval(interval);
  }, [isActive, time, step]);

  const formatTime = (time) => (time < 10 ? `0${time}` : time);

  const getTimeInMin = (time) => {
    let hours = Math.floor(time / 60 / 60);
    let minutes = Math.floor((time / 60) % 60);
    let seconds = time % 60;

    seconds = formatTime(seconds);
    minutes = formatTime(minutes);
    hours = formatTime(hours);

    return `${hours}:${minutes}:${seconds}`;
  };

  return (
    <>
      <h1 className={`title title--${theme}`}>Time: {getTimeInMin(time)}</h1>
      <div
        className="timer-progress"
        style={{ width: `${(time / initialTime) * 100}%` }}
      ></div>
    </>
  );
};

TimeDisplay.propTypes = {
  item: PropTypes.shape({
    initialTime: PropTypes.number,
    step: PropTypes.number,
    isActive: PropTypes.bool,
  }),
  formatTime: PropTypes.func,
  getTimeInMin: PropTypes.func,
};

export default TimeDisplay;
