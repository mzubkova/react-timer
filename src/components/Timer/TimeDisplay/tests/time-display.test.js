import React from "react";
import { shallow, mount } from "enzyme";
import TimeDisplay from "../TimeDisplay";

describe("TimeDisplay", () => {
  it("Should match shapshot", () => {
    const component = shallow(<TimeDisplay />);
    expect(component.html()).toMatchSnapshot();
  });
  it("Should have TimeDisplay", () => {
    const component = mount(<TimeDisplay />);
    expect(component.find("TimeDisplay")).toHaveLength(1);
  });
  it("Should have h1", () => {
    const component = mount(<TimeDisplay />);
    expect(component.find("h1")).toHaveLength(1);
  });
  it("Should have div", () => {
    const component = mount(<TimeDisplay />);
    expect(component.find("div")).toHaveLength(1);
  });
});
