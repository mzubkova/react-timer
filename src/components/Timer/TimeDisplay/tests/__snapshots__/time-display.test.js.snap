// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeDisplay Should match shapshot 1`] = `"<h1 class=\\"title title--light\\">Time: NaN:NaN:NaN</h1><div class=\\"timer-progress\\" style=\\"width:NaN%\\"></div>"`;
